kubelet: storage: teardown terminated pod volumes

sjenning opened the pull request:
This is a continuation of the work done in https://github.com/kubernetes/kubernetes/pull/36779

There really is no reason to keep volumes for terminated pods attached on the node.  This PR extends the removal of volumes on the node from memory-backed (the current policy) to all volumes.

@pmorie raised a concern an impact debugging volume related issues if terminated pod volumes are removed.  To address this issue, the PR adds a `--keep-terminated-pod-volumes` flag the kubelet and sets it for `hack/local-up-cluster.sh`.

For consideration in 1.6.

Fixes #35406

@derekwaynecarr @vishh @dashpole

```release-note
kubelet tears down pod volumes on pod termination rather than pod deletion
```

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/37228)
<!-- Reviewable:end -->

sjenning committed:
reclaim terminated pod volumes

smarterclayton commented:
In the long run I could imagine tagging specific containers and pods as having a "retainPolicy" that allows debugging of other aspects, not just volumes, post pod termination.  Things like logs, network artifacts, etc.  But that doesn't block or inhibit this work.

derekwaynecarr commented:
In the long run, i want to be able to schedule on local disk, so keeping disk cleanup with pod lifetime is important.  I guess we could do something like retainPolicy, but it would need to be visible to scheduler.

mikedanese commented:
@k8s-bot bazel test this, issue kubernetes/test-infra#1152

sjenning commented:
@saad-ali @kubernetes/sig-storage 

I would like to get this in early in the 1.6 dev cycle so it has time to soak.  Reviews appreciated!

vishh commented:
cc @dashpole who is working on cleaning up all volumes as part of pod termination.

saad-ali commented:
> I would like to get this in early in the 1.6 dev cycle so it has time to soak. Reviews appreciated!

Ack, will take a look as soon as 1.5 release is out.

derekwaynecarr commented:
@vishh -- you agree we could stage the work?  for example, accept this change before taking any larger refactor that @dashpole was looking at performing?

derekwaynecarr commented:
for example, it seems like we will need to have a flag like this to preserve legacy behavior no matter any future solution

vishh commented:
Staging is fine by me.

On Thu, Dec 1, 2016 at 2:36 PM, Derek Carr <notifications@github.com> wrote:

> for example, it seems like we will need to have a flag like this to
> preserve legacy behavior no matter any future solution
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/37228#issuecomment-264316964>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AGvIKFezPf1ySLm3jIy82uSFNlnoAEJ4ks5rD0vQgaJpZM4K4imT>
> .
>

sjenning commented:
@saad-ali can I get your review on this?  I'd like to give it maximum soak time in the 1.6 dev cycle if you approve of the change.

saad-ali commented:
> @saad-ali can I get your review on this? I'd like to give it maximum soak time in the 1.6 dev cycle if you approve of the change.

Reviewing

saad-ali commented on cmd/kubelet/app/options/options.go:
```diff
@@ -253,6 +253,7 @@ func (s *KubeletServer) AddFlags(fs *pflag.FlagSet) {
 	fs.BoolVar(&s.ExperimentalKernelMemcgNotification, "experimental-kernel-memcg-notification", s.ExperimentalKernelMemcgNotification, "If enabled, the kubelet will integrate with the kernel memcg notification to determine if memory eviction thresholds are crossed rather than polling.")
 	fs.Int32Var(&s.PodsPerCore, "pods-per-core", s.PodsPerCore, "Number of Pods per core that can run on this Kubelet. The total number of Pods on this Kubelet cannot exceed max-pods, so max-pods will be used if this calculation results in a larger number of Pods allowed on the Kubelet. A value of 0 disables this limit.")
 	fs.BoolVar(&s.ProtectKernelDefaults, "protect-kernel-defaults", s.ProtectKernelDefaults, "Default kubelet behaviour for kernel tuning. If set, kubelet errors if any of kernel tunables is different than kubelet defaults.")
+	fs.BoolVar(&s.KeepTerminatedPodVolumes, "keep-terminated-pod-volumes", s.KeepTerminatedPodVolumes, "Keep terminated pod volumes attached to the node afterh the pod terminates.  Can be useful for debugging volume related issues.")
```
What is the default for this flag, and where is it set?

sjenning replied:
The bool flags are defaulted to 0 automatically by Go afaict.  There are some bool flags that use pointers in v1alpha1 and make the distinction between true/false/unset cases.  But it this case, false is the desired default, and we don't care about tracking the whether the flag was explicitly set false or no present at all (same effect).

saad-ali commented on hack/local-up-cluster.sh:
```diff
@@ -537,6 +537,7 @@ function start_kubelet {
         --experimental-cgroups-per-qos=${EXPERIMENTAL_CGROUPS_PER_QOS} \
         --cgroup-driver=${CGROUP_DRIVER} \
         --cgroup-root=${CGROUP_ROOT} \
+        --keep-terminated-pod-volumes=true \
```
Why do this for local cluster up?

sjenning replied:
I figured `local-up-cluster.sh` is used for deploying dev environments where it is more likely the user will want to keep the volumes around for debug purposes.  I know that is what I would want personally.

saad-ali commented on pkg/apis/componentconfig/types.go:
```diff
@@ -483,6 +483,9 @@ type KubeletConfiguration struct {
 	// (binaries, etc.) to mount the volume are available on the underlying node. If the check is enabled
 	// and fails the mount operation fails.
 	ExperimentalCheckNodeCapabilitiesBeforeMount bool
+	// This flag, if set, instructs the kubelet to keep volumes from terminated pods attached to the node.
```
`attached` -> `mounted`
Ditto in other files

saad-ali reviewed (commented):
A few minor comments. Otherwise LGTM

jingxu97 commented:
@sjenning #36779 addressed the pod termination case on kubelet side, should we also do the same on master attach_detach_controller?

sjenning commented:
@saad-ali I think I addressed your feedback in the latest commits.  Can you look again?

@jingxu97 This is the first I'm learning of a controller the cares about volumes.  I see from https://github.com/kubernetes/kubernetes/issues/20262 this was created to mount cloud volumes (GPD, EBS, Cinder, etc) to the nodes.  Looks like a rough copy of the kubelet volume manager.  Not sure how the states of the world sync between the two.

Does anything need to be done wrt the attachdeatch controller?

saad-ali commented:
/lgtm

saad-ali commented:
Feel free to readd LGTM label after rebase. And please add a release note.

sjenning commented:
can I re-add an lgtm on my own PR?

/lgtm

vishh commented:
/lgtm

On Fri, Jan 20, 2017 at 11:11 AM, k8s-ci-robot <notifications@github.com>
wrote:

> @sjenning <https://github.com/sjenning>: you can't LGTM your own PR.
>
> In response to this comment
> <https://github.com/kubernetes/kubernetes/pull/37228#issuecomment-274153904>
> :
>
> can I re-add an lgtm on my own PR?
>
> /lgtm
>
> Instructions for interacting with me using PR comments are available here
> <https://github.com/kubernetes/community/blob/master/contributors/devel/pull-request-commands.md>.
> If you have questions or suggestions related to my behavior, please file an
> issue against the kubernetes/test-infra
> <https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:>
> repository. I understand the commands that are listed here
> <https://github.com/kubernetes/test-infra/blob/master/prow/commands.md>.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/37228#issuecomment-274153906>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AGvIKI1kkPryIObeg0H2Lyuy_lc7Nv5mks5rUQbFgaJpZM4K4imT>
> .
>

derekwaynecarr commented:
/lgtm
